### PR TITLE
WIP: IPv4/IPv6 dual stack support in kube-proxy (part of "dual stack phase 2")

### PR DESCRIPTION
**Pull request by vllry**
This PR should be merged with/onto #79386 - it's presented as-such for review/tracking purposes.

**Reviewer notes**
Commits marked `WIP` are not ready to review, other commits *probably* are.

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allows kube-proxy to run in IPv4 and IPv6 modes at the same time.

Main KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -1388,10 +1388,6 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
 			IP: ip,
 		})
 	}
-
-	if len(apiPodStatus.PodIPs) > 0 {
```
Comment (aojea) [inline pkg/kubelet/kubelet_pods.go]: I´m not familiar with this code but ... is this change related to this commit and kubenet?

File: pkg/proxy/apis/config/v1alpha2/defaults.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/defaults.go]: nit: 2019

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (vllry) [reply pkg/kubelet/kubelet_pods.go]: Unsure... I'm suspicious that it's an accident from a rebase, because I definitely didn't add it, and I can't find it in the parent branch https://github.com/kubernetes/kubernetes/pull/79386/files

File: pkg/proxy/apis/config/v1alpha2/defaults.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// getDefaultAddresses returns default address of healthz and metrics server
+// based on the given bind address. IPv6 addresses are enclosed in square
+// brackets for appending port.
+func getDefaultAddresses(bindAddress string) (defaultHealthzAddress, defaultMetricsAddress string) {
+	if net.ParseIP(bindAddress).To4() != nil {
+		return "0.0.0.0", "127.0.0.1"
+	}
+	return "[::]", "[::1]"
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/defaults.go]: question, do we need to use square brackets for the health and metrics addressed using IPv6 ?
if affirmative, should be better return the address without square brackets and use https://golang.org/pkg/net/#JoinHostPort to create the URL?

File: pkg/proxy/apis/config/v1alpha2/doc.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/doc.go]: nit s/2017/2019

File: pkg/proxy/apis/config/v1alpha2/doc.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=k8s.io/kubernetes/pkg/proxy/apis/config
+// +k8s:conversion-gen-external-types=k8s.io/kube-proxy/config/v1alpha1
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/doc.go]: I´m not very familiar with the API generation, but should this be v1alpha2?

File: pkg/proxy/apis/config/v1alpha2/doc.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+// +k8s:defaulter-gen=TypeMeta
+// +k8s:defaulter-gen-input=../../../../../vendor/k8s.io/kube-proxy/config/v1alpha1
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/doc.go]: v1alpha2?

File: pkg/proxy/apis/config/v1alpha2/register.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/register.go]: nit, s/2015/2019/

File: pkg/proxy/apis/config/v1alpha2/register.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubeproxyconfigv1alpha2 "k8s.io/kube-proxy/config/v1alpha2"
+)
+
+// GroupName is the group name used in this package
+const GroupName = "kubeproxy.config.k8s.io"
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha1"}
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/register.go]: v1alpha2?

File: pkg/proxy/apis/config/v1alpha2/defaults.go (hunk 3 above)
Comment (vllry) [reply pkg/proxy/apis/config/v1alpha2/defaults.go]: I think so - I copied that from the v1alpha1, but there may not have been a strong reason for it in the first place. Though looking at that, that I also need to revisit the logic... it should be checking what the bindAddresses are.

File: pkg/proxy/apis/config/validation/validation_test.go [hunk 9]
```diff
@@ -252,10 +271,10 @@ func TestValidateKubeProxyConfiguration(t *testing.T) {
 		},
 		{
 			config: kubeproxyconfig.KubeProxyConfiguration{
-				BindAddress:        "192.168.59.103",
+				BindAddress:        []string{"192.168.59.103"},
 				HealthzBindAddress: "0.0.0.0:10256",
 				MetricsBindAddress: "127.0.0.1:10249",
-				ClusterCIDR:        "192.168.59.0/24",
+				ClusterCIDR:        []string{"192.168.59.0/24"},
```
Comment (aojea) [inline pkg/proxy/apis/config/validation/validation_test.go]: maybe we should add some validation with IPv6 and dual-stack cidrs

File: staging/src/k8s.io/kube-proxy/config/v1alpha2/doc.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (aojea) [inline staging/src/k8s.io/kube-proxy/config/v1alpha2/doc.go]: nit , copyright year

File: staging/src/k8s.io/kube-proxy/config/v1alpha2/register.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (aojea) [inline staging/src/k8s.io/kube-proxy/config/v1alpha2/register.go]: nit, copyright year

File: staging/src/k8s.io/kube-proxy/config/v1alpha2/register.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name used in this package
+const GroupName = "kubeproxy.config.k8s.io"
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha1"}
```
Comment (aojea) [inline staging/src/k8s.io/kube-proxy/config/v1alpha2/register.go]: v1alpha2?

File: staging/src/k8s.io/kube-proxy/config/v1alpha2/types.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (aojea) [inline staging/src/k8s.io/kube-proxy/config/v1alpha2/types.go]: 2019?

File: cmd/kube-proxy/app/server.go [hunk 14]
```diff
@@ -658,9 +674,13 @@ func getConntrackMax(config kubeproxyconfig.KubeProxyConntrackConfiguration) (in
...
-	encounteredError = iptables.CleanupLeftovers(s.IptInterface) || encounteredError
-	encounteredError = ipvs.CleanupLeftovers(s.IpvsInterface, s.IptInterface, s.IpsetInterface, s.CleanupIPVS) || encounteredError
+	var encounteredError bool
+	for _, iface := range []*utiliptables.Interface{s.Ipt4Interface, s.Ipt6Interface} {
+		encounteredError := userspace.CleanupLeftovers(*iface)
+		encounteredError = iptables.CleanupLeftovers(*iface) || encounteredError
+		encounteredError = ipvs.CleanupLeftovers(s.IpvsInterface, *iface, s.IpsetInterface, s.CleanupIPVS) || encounteredError
+	}
+
 	if encounteredError {
```
Comment (aojea) [inline cmd/kube-proxy/app/server.go]: should the if go inside the for loop or do we have to wait until we clean all the interfaces?

File: cmd/kube-proxy/app/server.go [hunk 15]
```diff
@@ -658,9 +674,13 @@ func getConntrackMax(config kubeproxyconfig.KubeProxyConntrackConfiguration) (in
 
 // CleanupAndExit remove iptables rules and exit if success return nil
 func (s *ProxyServer) CleanupAndExit() error {
-	encounteredError := userspace.CleanupLeftovers(s.IptInterface)
-	encounteredError = iptables.CleanupLeftovers(s.IptInterface) || encounteredError
-	encounteredError = ipvs.CleanupLeftovers(s.IpvsInterface, s.IptInterface, s.IpsetInterface, s.CleanupIPVS) || encounteredError
+	var encounteredError bool
+	for _, iface := range []*utiliptables.Interface{s.Ipt4Interface, s.Ipt6Interface} {
+		encounteredError := userspace.CleanupLeftovers(*iface)
```
Comment (aojea) [inline cmd/kube-proxy/app/server.go]: does this clean the `encounteredError` from previous iteration?

File: cmd/kube-proxy/app/server_others.go [hunk 16]
```diff
@@ -71,23 +71,40 @@ func newProxyServer(
...
 	var dbus utildbus.Interface
+	dbus = utildbus.New() // TODO(vllry) does this need to be unique per-iptables interface?
 
 	// Create a iptables utils.
-	execer := exec.New()
+	execer := exec.New() // TODO(vllry) does this need to be unique per-iptables interface?
+
+	// Create iptables interfaces.
+	// Creates an IPv4 and/or IPv6 interface, depending on the types of BindAddresses.
+	for _, bindAddress := range config.BindAddress {
```
Comment (aojea) [inline cmd/kube-proxy/app/server_others.go]: I think that the behavior will be that the last addresses in config.BindAddress will get the iptXInterface ... don´t remember if we are limiting config.BindAddress to be only IPv4, IPv6 or having one IPv4 and other IPv6 address

**Comment by vllry** [comment]
@aojea thanks so much for the review so far! I noticed that I had some commits that were removed from the parent branch, so I did some manual rebasing to match https://github.com/kubernetes/kubernetes/pull/79386/commits

Other than that, you exposed a bunch of things I skipped over in the api defaults/conversion. I'll go back and address those.

File: pkg/proxy/apis/config/validation/validation_test.go (hunk 9 above)
Comment (vllry) [reply pkg/proxy/apis/config/validation/validation_test.go]: I stuck in a case, I'll put in more later.

File: pkg/proxy/apis/config/v1alpha2/defaults.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	if obj.FeatureGates == nil {
+		obj.FeatureGates = make(map[string]bool)
+	}
+}
+
+// getDefaultAddresses returns default address of healthz and metrics server
+// based on the given bind address. IPv6 addresses are enclosed in square
+// brackets for appending port.
+// TODO(@vllry) should this have multiple health/metrics interfaces?
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/defaults.go]: that´s a good question

File: pkg/proxy/apis/config/validation/validation_test.go [hunk 18]
```diff
@@ -82,10 +82,29 @@ func TestValidateKubeProxyConfiguration(t *testing.T) {
...
+				Min:                   pointer.Int32Ptr(1),
+				TCPEstablishedTimeout: &metav1.Duration{Duration: 5 * time.Second},
+				TCPCloseWaitTimeout:   &metav1.Duration{Duration: 5 * time.Second},
+			},
+		},
+		{
+			BindAddress:        []string{"192.168.59.103", "fd12:3456:789a:1::1"},
+			HealthzBindAddress: "",
+			MetricsBindAddress: "127.0.0.1:10249",
+			ClusterCIDR:        []string{"192.168.59.0/24", "::/0"},
```
Comment (aojea) [inline pkg/proxy/apis/config/validation/validation_test.go]: `::/0`means default route, I guess we should use another CIDR, i.e. "2001:db:1::/64"

File: pkg/proxy/util/utils.go [hunk 19]
```diff
@@ -241,8 +241,5 @@ func AppendPortIfNeeded(addr string, port int32) string {
 	}
 
 	// Append port to address.
-	if ip.To4() != nil {
-		return fmt.Sprintf("%s:%d", addr, port)
-	}
-	return fmt.Sprintf("[%s]:%d", addr, port)
+	return net.JoinHostPort(addr, string(port))
```
Comment (aojea) [inline pkg/proxy/util/utils.go]: do you know if we have to change `pkg/proxy/apis/config/v1alpha1/defaults.go` to not return the ipv6 default address with square brackets?

File: pkg/proxy/apis/config/v1alpha2/defaults.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+	if obj.ClientConnection.Burst == 0 {
+		obj.ClientConnection.Burst = 10
+	}
+	if obj.FeatureGates == nil {
+		obj.FeatureGates = make(map[string]bool)
+	}
+}
+
+// getDefaultAddresses returns default address of healthz and metrics server
+// based on the given bind address. IPv6 addresses are enclosed in square
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/defaults.go]: is the comment 
> IPv6 addresses are enclosed in square
correct?

File: pkg/proxy/apis/config/v1alpha2/defaults.go [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+// brackets for appending port.
+// TODO(@vllry) should this have multiple health/metrics interfaces?
+func getDefaultAddresses(bindAddress string) (defaultHealthzAddress, defaultMetricsAddress string) {
+	if net.ParseIP(bindAddress).To4() != nil {
```
Comment (aojea) [inline pkg/proxy/apis/config/v1alpha2/defaults.go]: this will return nil if the bindAddress is not valid, right?

File: cmd/kube-proxy/app/server_others.go [hunk 22]
```diff
@@ -50,6 +50,28 @@ import (
...
+// (Doesn't check that IP addresses are valid)
+// Returns: []{<ipv4 || ipv6>} || []{<ipv4, ipv6>} || []{<ipv6, ipv4>}
+func parseBindAddresses(addresses []string) []string {
+	filteredAddresses := make([]string, 0)
+
+	hasV4 := false
+	hasV6 := false
+
+	for _, addr := range addresses {
+		if !hasV4 && net.ParseIP(addr).To4() != nil {
```
Comment (khenidak) [inline cmd/kube-proxy/app/server_others.go]: You probably will need to do that in a lot of places. Let us use https://github.com/kubernetes/utils/blob/master/net/net.go#L113 it is already vendored in k/k

File: cmd/kube-proxy/app/server_others.go [hunk 23]
```diff
@@ -80,25 +102,29 @@ func newProxyServer(
 	var kernelHandler ipvs.KernelHandler
 	var ipsetInterface utilipset.Interface
 
-	var dbus utildbus.Interface
-	dbus = utildbus.New() // TODO(vllry) does this need to be unique per-iptables interface?
+	var dbus4 utildbus.Interface
+	dbus4 = utildbus.New()
+	var dbus6 utildbus.Interface
+	dbus6 = utildbus.New()
```
Comment (khenidak) [inline cmd/kube-proxy/app/server_others.go]: do we really need a different bus for each `ipt` interface?

File: cmd/kube-proxy/app/server_others.go [hunk 24]
```diff
@@ -80,25 +102,29 @@ func newProxyServer(
...
 	// Create a iptables utils.
-	execer := exec.New() // TODO(vllry) does this need to be unique per-iptables interface?
+	execer := exec.New()
+
+	filteredBindAddresses := parseBindAddresses(config.BindAddress)
 
 	// Create iptables interfaces.
 	// Creates an IPv4 and/or IPv6 interface, depending on the types of BindAddresses.
-	for _, bindAddress := range config.BindAddress {
+	for _, bindAddress := range filteredBindAddresses {
```
Comment (khenidak) [inline cmd/kube-proxy/app/server_others.go]: same as before, use common `k/util` package.

File: cmd/kube-proxy/app/server_others.go [hunk 25]
```diff
@@ -50,6 +50,28 @@ import (
...
+// Reduces a list of bindaddresses down to the first ipv4 and first ipv6 address in the list.
+// (Doesn't check that IP addresses are valid)
+// Returns: []{<ipv4 || ipv6>} || []{<ipv4, ipv6>} || []{<ipv6, ipv4>}
+func parseBindAddresses(addresses []string) []string {
+	filteredAddresses := make([]string, 0)
+
+	hasV4 := false
+	hasV6 := false
+
+	for _, addr := range addresses {
```
Comment (khenidak) [inline cmd/kube-proxy/app/server_others.go]: if we are to do that, then we need to log to user that we ignored (and why) something they provided.

**Comment by vllry** [comment]
I’ve been pretty inconsistently making changes as I move around the codebase - tomorrow I’ll try to polish the API changes and solidify the config changes.

The proxier will definitely be up on the air for a few more days, it’s main loop is a beast and I’m still not sure if I should support both stacks in the same sync loop, or have 2 proxiers.

**Comment by uablrek** [comment]
@vllry 

I would like to try this PR locally and have two questions;

I have cloned your branch and built successfully, but I wonder if PR #79386 is "included"?
```
git clone --depth 1 -b dualstack-phase2-kubeproxy git@github.com:vllry/kubernetes.git
```

I also get a crash in `kube-proxy` which I think is a configuration problem. At first I use my old kube-proxy config without modifications. Is there an updated kube-proxy config file example I can use as template?

The crash;
```
panic: runtime error: index out of range

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime.Convert_Slice_string_To_string(...)
        /home/uablrek/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/conversion.go:68
...
k8s.io/kubernetes/cmd/kube-proxy/app.(*Options).ApplyDefaults(0xc00002a000, 0xc000108540, 0xc000274060, 0x3, 0x197b740)
        /home/uablrek/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:415 +0x8f
k8s.io/kubernetes/cmd/kube-proxy/app.NewProxyCommand(0xc00003a1b0)
        /home/uablrek/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:464 +0x1b9
main.main()
        _output/local/go/src/k8s.io/kubernetes/cmd/kube-proxy/proxy.go:37 +0x90
```

Version;
```
# kubectl version
Client Version: version.Info{Major:"", Minor:"", GitVersion:"v0.0.0-master+$Format:%h$", GitCommit:"d33a7ac0a7d5127934cf14e9bae4fec6928ce373", GitTreeState:"dirty", BuildDate:"2019-07-22T09:12:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```

File: pkg/proxy/util/utils.go (hunk 19 above)
Comment (vllry) [reply pkg/proxy/util/utils.go]: Look like this line is failing (produces a garbled port)

**Comment by uablrek** [comment]
Please consider creating instances of the current proxier, one for ipv4 and one for ipv6. I think it vill be simpler for iptables and ipvs proxy-modes, see https://github.com/kubernetes/kubernetes/pull/80627.

A large benefit is that the current proxier, which is widely used and very stable, can be invoked without any large alteratios if the used does not set the feature gateway. This minimize the risk for problems when dual-stack is introduced in alpha and beta.

For a future proxy-modes based on "nft" it may be better with just one proxier since nft allows a mix of ipv4 anf ipv6, so both approaches should be possible.

There may still be problems that makes it impossible to run two proxiers in parallel but IMHO those problems should be fixed if they exist rather than enforcing a single proxier. For instance I don't know how the "recorder" works.

**Comment by vllry** [comment]
> Please consider creating instances of the current proxier, one for ipv4 and one for ipv6. I think it vill be simpler for iptables and ipvs proxy-modes, see #80627.
> 
> A large benefit is that the current proxier, which is widely used and very stable, can be invoked without any large alteratios if the user does not set the dual-stack feature gateway. This minimize the risk for problems when dual-stack is introduced in alpha and beta.
> 
> For a future proxy-modes based on "nft" it may be better with just one proxier since nft allows a mix of ipv4 anf ipv6, so both approaches should be possible.
> 
> There may still be problems that makes it impossible to run two proxiers in parallel but IMHO those problems should be fixed if they exist rather than enforcing a single proxier. For instance I don't know how the "recorder" works.

I specifically avoided this approach, out of concern that ipv4 and ipv6 rules could get out of sync with one another. It's not an order-of-magnitude thing, but it's also less efficient to independently be polling endpoints.

@thockin any strong opinions on this?

**Comment by uablrek** [comment]
I do not understand the "out of sync" concern. Ip6tables and iptables rules will always handle different services and different endpoints, so they never be in sync.

The endpoint (and service) polling is still done at one place, but the resulting function calls over the proxier API are dispatched to different proxier instances depending on address family.

**Review by thockin — COMMENTED** [review]
I think you have the right config stuff here to get multiple CIDRs in, but have you looked at #80627  ?  I really like how clean and low-impact that is.  Can we blend them?

File: pkg/proxy/apis/config/validation/validation_test.go (hunk 18 above)
Comment (dcbw) [reply pkg/proxy/apis/config/validation/validation_test.go]: @aojea @vllry yeah need to pick a CIDR in some unroutable range and use it; perhaps the "site-local" v6 range of FEC0::/10. Just make something up in there.

**Comment by vllry** [comment]
I'm trying to keep track of all the parallel work going on.

After going over the code more, I think the meta-proxier approach from 80627 makes more sense, so I'll look at implementing that. The current iptables proxier code in this PR is logically simple, but in practice it's spaghetti.

I noticed an amount over overlapping code while rebasing this PR - not totally sure how much of the config or proxier handling is already implemented elsewhere.

**Comment by vllry** [comment]
I've scrubbed most of the proxier changes, going to go through the IPVS proxier / metaproxier stuff.

**Comment by vllry** [comment]
Closing PR, replacing with #82462 (iptables proxier)
